src/psbtv2.ts

```diff
@@ -0,0 +1,910 @@
+// Values can be of various different types or formats. Here we leave them as
+// Buffers so that getters can decide how they should be formatted.
+type Value = Buffer
+
+
+// These keytypes are hex bytes, but here they are used as string enums to
+// assist in Map lookups. See type Key above for more info.
+enum KeyType {
+  PSBT_GLOBAL_XPUB = '01',
+  PSBT_GLOBAL_TX_VERSION = '02',
+  PSBT_GLOBAL_FALLBACK_LOCKTIME = '03',
+  PSBT_GLOBAL_INPUT_COUNT = '04',
+  PSBT_GLOBAL_OUTPUT_COUNT = '05',
+  PSBT_GLOBAL_TX_MODIFIABLE = '06',
+  PSBT_GLOBAL_VERSION = 'fb',
+  PSBT_GLOBAL_PROPRIETARY = 'fc',
+
+  PSBT_IN_NON_WITNESS_UTXO = '00',
+  PSBT_IN_WITNESS_UTXO = '01',
+  PSBT_IN_PARTIAL_SIG = '02',
+  PSBT_IN_SIGHASH_TYPE = '03',
+  PSBT_IN_REDEEM_SCRIPT = '04',
+  PSBT_IN_WITNESS_SCRIPT = '05',
+  PSBT_IN_BIP32_DERIVATION = '06',
+  PSBT_IN_FINAL_SCRIPTSIG = '07',
+  PSBT_IN_FINAL_SCRIPTWITNESS = '08',
+  PSBT_IN_POR_COMMITMENT = '09',
+  PSBT_IN_RIPEMD160 = '0a',
+  PSBT_IN_SHA256 = '0b',
+  PSBT_IN_HASH160 = '0c',
+  PSBT_IN_HASH256 = '0d',
+  PSBT_IN_PREVIOUS_TXID = '0e',
+  PSBT_IN_OUTPUT_INDEX = '0f',
+  PSBT_IN_SEQUENCE = '10',
+  PSBT_IN_REQUIRED_TIME_LOCKTIME = '11',
+  PSBT_IN_REQUIRED_HEIGHT_LOCKTIME = '12',
+  PSBT_IN_TAP_KEY_SIG = '13',
+  PSBT_IN_TAP_SCRIPT_SIG = '14',
+  PSBT_IN_TAP_LEAF_SCRIPT = '15',
+  PSBT_IN_TAP_BIP32_DERIVATION = '16',
+  PSBT_IN_TAP_INTERNAL_KEY = '17',
+  PSBT_IN_TAP_MERKLE_ROOT = '18',
+  PSBT_IN_PROPRIETARY = 'fc',
+  
+  PSBT_OUT_REDEEM_SCRIPT = '00',
+  PSBT_OUT_WITNESS_SCRIPT = '01',
+  PSBT_OUT_BIP32_DERIVATION = '02',
+  PSBT_OUT_AMOUNT = '03',
+  PSBT_OUT_SCRIPT = '04',
+  PSBT_OUT_TAP_INTERNAL_KEY = '05',
+  PSBT_OUT_TAP_TREE = '06',
+  PSBT_OUT_TAP_BIP32_DERIVATION = '07',
+  PSBT_OUT_PROPRIETARY = 'fc',
+}
+
+
+// Provided to friendly-format the PSBT_GLOBAL_TX_MODIFIABLE bitmask from
+// PsbtV2.PSBT_GLOBAL_TX_MODIFIABLE which returns PsbtGlobalTxModifiableBits[]. 
+enum PsbtGlobalTxModifiableBits {
+  INPUTS = 'INPUTS', // 0b000000001
+  OUTPUTS = 'OUTPUTS', // 0b000000010
+  SIGHASH_SINGLE = 'SIGHASH_SINGLE', // 0b000000100
+}
+
+
+/* 
+Global Constants
+ */
+
+const PSBT_MAGIC_BYTES = Buffer.from([0x70, 0x73, 0x62, 0x74, 0xff]);
+const PSBT_MAP_SEPARATOR = Buffer.from([0x00]);
+const BIP_32_NODE_REGEX = /(\/[0-9]+'?)/ig;
+const ERROR_PSBT_NOT_VALID = Error("Not a valid psbt");
+
+
+/* 
+Helper Functions
+*/
+
+// Return psbt as Buffer.
+function bufferize (psbt:string|Buffer):Buffer {
+  let buf:Buffer|undefined;
+
+  if (Buffer.isBuffer(psbt)) {
+    buf = psbt;
+  } else if (typeof psbt === 'string' && (/^([0-9A-Fa-f][0-9A-Fa-f])+$/).test(psbt)) {
+    // Is a hex string
+    buf = Buffer.from(psbt, 'hex');
+  } else if (
+    typeof psbt === 'string' && 
+    (/^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=|[A-Za-z0-9+/]{4})$/).test(psbt)
+  ) {
+    // Is a base64 string
+    buf = Buffer.from(psbt, 'base64');
+  } else {
+    throw ERROR_PSBT_NOT_VALID;
+  }
+
+  return buf;
+}
```

Comment:
Perhaps this could use some better structure, but I don't know that everything in your suggestion is better. The base64 check I committed does a more robust check of valid formats. During development, I was running into cases where it was switching to the wrong format case. The regex, though long, ensures valid padding and chunking of base64. Likewise, the hex regex is checking for full bytes. If a byte is a nibble nothing will work, so it with throw.

Reply:
Yeah I might have hosed the regex, main concern is changing to an early return  and cutting out multiple if typeof psbt === "string" checks if possible for the sake of readability.

Reply:
Done. I usually try to avoid a multiple return kind of flow, but this function is short enough.